### PR TITLE
fix(menu/save-project): Add missing XML prolog on document save

### PR DESCRIPTION
**Pull request by danyill**
Most of the DOM methods for detecting the declaration/prolog are deprecated (`Document.xmlVersion`, `Document.xmlEncoding`, `DOMImplementation.hasFeature()`) so I avoided using them. There do not appear to be replacements.

As per previous discussions on Save Project, I don't think we can add a test without modifying the code for the test.

I think we should leave the issue open until we have a test for this as part of the Action/Editing API. i.e. carry out action, check that the document is well-formed.

I would be grateful for testing/review as this is quite important.

Part of #1163

### CODE REVIEW

**Comment by danyill** [comment]
Ooops wrong branch. Let me tidy this up.

**Review by ca-d — DISMISSED** [review]
LGTM, but is there a chance this might lead to prologue duplication in case there's a prologue with whitespace nodes before it or something (e.g. attribute order) is formatted differently? I don't know if these things can happen, or if there could ever be other encoding attribute values e.g.

File: src/menu/SaveProject.ts [hunk 1]
```diff
@@ -19,8 +19,15 @@ export default class SaveProjectPlugin extends LitElement {
 
   async run(): Promise<void> {
     if (this.doc) {
+      const xmlDeclaration = '<?xml version="1.0" encoding="UTF-8"?>'
+      let documentAsString = formatXml(new XMLSerializer().serializeToString(this.doc))
+
+      // Add XML declaration/prolog if it's been stripped
+      // TODO: This can be removed once the improved OpenSCD core edit API is present
+      documentAsString = documentAsString.startsWith(xmlDeclaration) ? documentAsString : xmlDeclaration + "\n" + documentAsString
```
[the remark this answers is not in the document]
Comment (danyill) [reply src/menu/SaveProject.ts:27]: I think it is low risk, but let's do that. I think other encodings are turned into UTF-8 anyway.

**Comment by danyill** [comment]
> LGTM, but is there a chance this might lead to prologue duplication in case there's a prologue with whitespace nodes before it or something (e.g. attribute order) is formatted differently? I don't know if these things can happen, or if there could ever be other encoding attribute values e.g.

I believe at best the document which did this would not be well-formed which is a reasonable requirement (?).

The spec requires that to be well formed the prolog should be unambiguously the first item (after the byte order mark): https://www.w3.org/TR/xml/#sec-prolog-dtd see section 2.1

File: src/menu/SaveProject.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (danyill) [reply src/menu/SaveProject.ts]: > I think it is low risk, but let's do that. I think other encodings are turned into UTF-8 anyway.

I'm not sure this is true but I tried a few different encodings and couldn't produce a problem... In any case I've never seen a tool or SCL document in anything other than UTF-8

**Comment by danyill** [comment]
Incidentally, it amuses me that we didn't notice this earlier. Anyone with a file editor GUI which inspected the files might have shown something like this:

image

In due course, I'll do a separate PR to tidy these and likely others.

**Review by ca-d — APPROVED** [review]
Sorry for not re-approving that sooner, LGTM.

**Pull request merged**